kubernetes/gaffer-jhub/files/tests/notebook_tests.py

```diff
@@ -0,0 +1,166 @@
+			return True
+		time.sleep(1)
+
+	raise Exception('Timed out waiting for notebook server to be deleted: {}/{}'.format(username, servername))
+
+def exec_command_in_pod(pod_name, namespace, cmd):
+	api = kubernetes.client.api.core_v1_api.CoreV1Api()
+
+	exec_command = ['/bin/bash', '-c', cmd]
+	response = kubernetes.stream.stream(
+		api.connect_get_namespaced_pod_exec,
+		pod_name,
+		namespace,
+		command=exec_command,
+		stderr=True,
+		stdin=False,
+		stdout=True,
+		tty=False,
+		_preload_content=False
+	)
+
+	response.run_forever()
+	return_code = response.returncode
+	output = response.read_all()
+
+	print(output)
+	print('ReturnCode:', return_code)
+
+	return return_code
+
+def test_custom_kubespawner_is_being_used():
+	info = get_jhub_info()
+	assert info['spawner']['class'] == 'builtins.KubeSpawnerWithForwardOptionsServer'
+
+def test_jhub_profiles(username, servername, namespace, service_account_name):
+	ensure_user_exists(username)
+
+	options = get_notebook_spawn_options(username, servername)
+	for profile_slug in options['profiles']:
+		profile = options['profiles'][profile_slug]
+		print('')
+		print('=== Testing profile {}: {}'.format(profile['slug'], profile['display_name']))
+		print(profile)
+
+		if namespace not in options['namespaces']:
+			raise Exception('Unable to spawn any notebooks as user {} does not have permissions to the {} namespace!'.format(username, namespace))
+
+		service_account_id = None
+		for service_account in options['namespaces'][namespace]['serviceAccounts']:
+			if service_account['name'] == service_account_name:
+				service_account_id = service_account['id']
+				break
+		if not service_account_id:
+			raise Exception('Unable to spawn any notebooks as the service account {} is not available to user {} in the {} namespace!'.format(service_account_name, username, namespace))
+
+		server_options = {
+			'profile': [profile['slug']],
+			'k8s_namespace': [namespace],
+			'k8s_service_account': [service_account_id],
+			'resources_mem': [1024]
+		}
+		if 'hdfs' in options and len(options['hdfs']) > 0:
+			server_options['hdfs'] = [options['hdfs'][0]['id']]
+		if 'graphs' in options and len(options['graphs']) > 0:
+			server_options['gaffer_graph'] = [options['graphs'][0]['id']]
+
+		try:
+			server = spawn_user_notebook(username, servername, server_options)
+
+			cmd = 'set -e; if [ -d /examples ]; then for file in /examples/*.ipynb; do echo Executing ${file}...; jupyter nbconvert ${file} --to python --stdout --execute --debug; done fi'
+			rc = exec_command_in_pod(server['state']['pod_name'], namespace, cmd)
+			assert rc == 0
+
+		finally:
+			delete_user_notebook(username, servername)
+
+test_custom_kubespawner_is_being_used()
+
+# E2E Test
+# - Query options server and get a list of all available profiles and options
+# - Spawn a Jupyter server for each profile
+# - Check that all example notebooks run successfully
+test_jhub_profiles(USERNAME, SERVERNAME, NAMESPACE, SERVICE_ACCOUNT_NAME)
```

Comment:
Are these tests run as part of the helm test jhub?

Reply:
Yep, `helm test` deploys the `gaffer-jhub-test-notebooks` pod which mounts and executes this script.

Reply:
I presumed it worked this way just wanted to double check, cheers!